File: collectous/app/src/main/java/org/anuhisoc/collectous/entry/SignInFragment.java
package org.anuhisoc.collectous.entry;

import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.activity.result.ActivityResult;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts.StartActivityForResult;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;

import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.common.api.ApiException;
import com.google.android.gms.tasks.Task;
import com.google.android.material.snackbar.BaseTransientBottomBar;
import com.google.android.material.snackbar.Snackbar;

import org.anuhisoc.collectous.databinding.FragmentSignInBinding;

import timber.log.Timber;



public class SignInFragment extends Fragment {


    private FragmentSignInBinding binding;
    private EntryViewModel entryViewModel;
    private GoogleSignInClient googleSignInClient;

    public SignInFragment() { }

    private final ActivityResultLauncher<Intent> signInLauncherHandler =
            registerForActivityResult(new StartActivityForResult(), result -> {
                Timber.d("Sign In Activity Result %s", result.toString());
                handleSignInResult(result);
            });


    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        binding = FragmentSignInBinding.inflate(inflater,container,false);
        return binding.getRoot();
    }


    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if(getActivity()!=null)
            entryViewModel = new ViewModelProvider(getActivity()).get(EntryViewModel.class);
        GoogleSignInOptions signInOption = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestEmail()
                .requestProfile()
                .build();

        if(getContext()!=null)
            googleSignInClient = GoogleSignIn.getClient(getContext(),signInOption);
    }


    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        binding.signInButton.setOnClickListener(signInButton -> signIn());
    }


    private void signIn(){
        Intent signInIntent = googleSignInClient.getSignInIntent();
        signInLauncherHandler.launch(signInIntent);
    }


    private void handleSignInResult(ActivityResult result) {
        try {
            Task<GoogleSignInAccount> task = GoogleSignIn.getSignedInAccountFromIntent(result.getData());
            Timber.d("Successful sign in ");
            GoogleSignInAccount account = task.getResult(ApiException.class);

            Snackbar snackbar = Snackbar.make(binding.getRoot(),"Successfully Signed In", BaseTransientBottomBar.LENGTH_SHORT);

            snackbar.addCallback(new BaseTransientBottomBar.BaseCallback<Snackbar>() {
                @Override
                public void onDismissed(Snackbar transientBottomBar, int event) {
                    super.onDismissed(transientBottomBar, event);
                    entryViewModel.updateGoogleAccount(account);
                    snackbar.removeCallback(this);
                }
            });
            snackbar.show();

        } catch (ApiException e) {
            Timber.d("Failed sign in ");
            Snackbar.make(binding.getRoot(),"Error",Snackbar.LENGTH_SHORT).show();
        }
    }

}

File: collectous/app/src/main/java/org/anuhisoc/collectous/entry/SignInFragment.kt
package org.anuhisoc.collectous.entry

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.activity.result.ActivityResult
import androidx.activity.result.contract.ActivityResultContracts.StartActivityForResult
import androidx.core.os.bundleOf
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Observer
import androidx.lifecycle.lifecycleScope
import androidx.navigation.fragment.findNavController
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.common.api.ApiException
import com.google.android.gms.common.api.Status
import com.google.android.material.snackbar.BaseTransientBottomBar
import com.google.android.material.snackbar.BaseTransientBottomBar.BaseCallback
import com.google.android.material.snackbar.Snackbar
import kotlinx.coroutines.launch
import org.anuhisoc.collectous.R
import org.anuhisoc.collectous.databinding.FragmentSignInBinding
import timber.log.Timber

class SignInFragment : Fragment() {
    private lateinit var binding: FragmentSignInBinding

    private val signInViewModel: SignInViewModel by viewModels()
    private val signInLauncherHandler = registerForActivityResult(StartActivityForResult()) { result: ActivityResult ->
        Timber.d("Sign In Activity Result %s", result.toString())
        handleSignInResult(result)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View {
        binding = FragmentSignInBinding.inflate(inflater, container, false)
        return binding.root
    }


    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.signInButton.setOnClickListener { launchSignInProcess() }

        signInViewModel.isSuccessSnackBarDismissed.observe(viewLifecycleOwner, Observer {
            isSuccessSnackBarDismissed -> if(isSuccessSnackBarDismissed==true){
            findNavController().navigate(R.id.action_signInFragment_to_setup, bundleOf("name" to signInViewModel.accountName))
        } })

    }

    private fun launchSignInProcess() {
        Timber.d("signIn process launched")
        val signInIntent = signInViewModel.googleSignInClient.signInIntent
        signInLauncherHandler.launch(signInIntent)
    }


    private fun handleSignInResult(result: ActivityResult) {
        try {
            val task = GoogleSignIn.getSignedInAccountFromIntent(result.data)
            Timber.d("Successful sign in ")
            val account = task.getResult(ApiException::class.java)
            lifecycleScope.launch {
                val isAccUpdatedSuccessfully = signInViewModel.updateGoogleAccount(account)
                if(isAccUpdatedSuccessfully){
                    launchSignInSuccessFeedback()
                }else{
                    throw ApiException(Status(Status.RESULT_CANCELED.statusCode))
                }
            }
        } catch (e: ApiException) {
            Timber.d("Failed sign in %s", e.toString())
            Snackbar.make(binding.root, "Error", Snackbar.LENGTH_SHORT).show()
        }
    }


    private fun launchSignInSuccessFeedback(){
        val snackBar = Snackbar.make(binding.root, "Successfully signed in!", BaseTransientBottomBar.LENGTH_SHORT)
        snackBar.addCallback(object : BaseCallback<Snackbar?>() {
            override fun onDismissed(transientBottomBar: Snackbar?, event: Int) {
                super.onDismissed(transientBottomBar, event)
                signInViewModel.setSnackBarDismissed()
                snackBar.removeCallback(this)
            }
        })
        snackBar.show()
    }
}